javascript/ql/src/Statements/IgnoreConcatReturn.ql

```diff
@@ -0,0 +1,42 @@
+/**
+ * @name Ignoring return from concat
+ * @description The concat method does not modify an array, ignoring the result of a call to concat is therefore generally an error.
+ * @kind problem
+ * @problem.severity warning
+ * @id js/ignore-result-from-concat
+ * @tags maintainability,
+ *       correctness
+ * @precision high
+ */
+
+import javascript
+import Expressions.ExprHasNoEffect
+
+DataFlow::SourceNode array(DataFlow::TypeTracker t) {
+  t.start() and
+  result instanceof DataFlow::ArrayCreationNode
+  or
+  exists (DataFlow::TypeTracker t2 |
+    result = array(t2).track(t2, t)
+  )
+}
+
+DataFlow::SourceNode array() { result = array(DataFlow::TypeTracker::end()) }
+
+predicate isArrayMethod(DataFlow::MethodCallNode call) {
+  call.getReceiver().getALocalSource() = array()
+}
+
+predicate isIncomplete(DataFlow::Node node) {
+  any(DataFlow::Incompleteness cause | node.analyze().getAValue().isIndefinite(cause)) != "global"
+}
+
+from DataFlow::CallNode call
+where 
+  isArrayMethod(call) and
+  call.getCalleeName() = "concat" and
+  call.getNumArgument() = 1 and 
+  (call.getArgument(0).getALocalSource() = array() or isIncomplete(call.getArgument(0))) and 
+  not call.getArgument(0).asExpr().(ArrayExpr).getSize() = 0 and 
```

Comment:
What about the receiver being the empty array?

Reply:
The call does nothing if the input is literally the empty array, and I got tired of looking at [tests like these](https://lgtm.com/projects/g/binoc-graveyard/mozilla-esr24/snapshot/787bf918e25e292675d5977c105c825b0bf5fe47/files/js/src/jit-test/tests/auto-regress/bug528048.js?sort=name&dir=ASC&mode=heatmap#xafd0777a79296517:1).

Reply:
I'm not sure I understand your answer. I do agree with this whitelisting. I was just wondering about the symmetric case of `[].concat(xs)`, which I saw in a few tests.

Reply:
> Is this PR ready for an editorial review or shall I hold off until all the technical discussions have taken place?

Hold off for now.

Reply:
Oh. I misread your comment, sorry. I was thinking about the way I wrote the query and thought you were talking about the argument not the receiver. 

You have a good point with the receiver being the empty array. I think we should remove those.